File: pmsender/PMFormatter.java
package pmsender;

import twao.Player;
import twao.VillageAssignment;
import twao.World;
import twao.villages.Village;

import java.text.SimpleDateFormat;
import java.util.*;
import java.util.List;

public class PMFormatter {
    private final World world;

    private int previousDay = -1; //flag init
    private boolean nextDay = false; //flag init

    private static final int SECONDS_PER_MIN        = 60;
    private static final int GROUP_SIZE             = 5;
    private static final String OPEN_SPOILER        = "[spoiler]\n";
    private static final String CLOSE_SPOILER       = "[/spoiler]\n\n\n";

    ResourceBundle bundle = ResourceBundle.getBundle("resources.bundles.pmformatter");

    private final String REQUIREMENTS_HEADER = bundle.getString("REQUIREMENTS_HEADER");
    private final String NOBLE_HEADER        = bundle.getString("NOBLE_HEADER");
    private final String OFF_HEADER          = bundle.getString("OFF_HEADER");
    private final String FAKE_HEADER         = bundle.getString("FAKE_HEADER");
    private final String FAKENOBLE_HEADER    = bundle.getString("FAKENOBLE_HEADER");
    private final String EXECUTION_TEXT      = bundle.getString("EXECUTION_TEXT");

    private enum Unit {
        RAM,
        NOBLE
    }

    public PMFormatter(World world) {
        this.world = world;
    }

    public String get(Player player) {
        StringBuilder sbuilder = new StringBuilder();

        if (player.getNobleAssignments().size() > 0 || player.getFakeNobleAssignments().size() > 0) {
            sbuilder.append(REQUIREMENTS_HEADER);
            generateNobleRequirements(sbuilder, player);
        }

        if (player.getNobleAssignments().size() > 0) {
            sortAssignmentsList(player.getNobleAssignments());

            sbuilder.append(NOBLE_HEADER);
            generateCommandsList(sbuilder, player.getNobleAssignments(), getUnitSpeed(Unit.NOBLE));
        }

        if (player.getOffAssignments().size() > 0) {
            sortAssignmentsList(player.getOffAssignments());

            sbuilder.append(OFF_HEADER);
            generateCommandsList(sbuilder, player.getOffAssignments(), getUnitSpeed(Unit.RAM));
        }

        if (player.getFakeAssignments().size() > 0) {
            sortAssignmentsList(player.getFakeAssignments());

            sbuilder.append(FAKE_HEADER);
            generateCommandsList(sbuilder, player.getFakeAssignments(), getUnitSpeed(Unit.RAM));
        }

        if (player.getFakeNobleAssignments().size() > 0) {
            sortAssignmentsList(player.getFakeNobleAssignments());

            sbuilder.append(FAKENOBLE_HEADER);
            generateCommandsList(sbuilder, player.getFakeNobleAssignments(), getUnitSpeed(Unit.NOBLE));
        }

        return sbuilder.toString();
    }

    private void generateNobleRequirements(StringBuilder sbuilder, Player player) {
        HashMap<Village, Integer> requirements = new HashMap<>();

        List<VillageAssignment> nobleAssignments = new LinkedList<>();
        nobleAssignments.addAll(player.getNobleAssignments());
        nobleAssignments.addAll(player.getFakeNobleAssignments());

        sbuilder.append(OPEN_SPOILER);

        int previousValue;
        for (VillageAssignment a : nobleAssignments) {
            if (requirements.get(a.getDeparture()) == null) {
                requirements.put(a.getDeparture(), 1);
            } else {
                previousValue = requirements.get(a.getDeparture());
                requirements.put(a.getDeparture(), ++previousValue);
            }
        }

        // {1: [Village, Village], 2: [], 3: [], ... }
        HashMap<Integer, List<Village>> sortedMap = sortRequirements(requirements);

        int iterator = 0;
        for (int value : sortedMap.keySet()) {
            for (Village vil : sortedMap.get(value)) {
                sbuilder.append(String.format("%d || %s\n", value, vil.toString()));

                if (++iterator % GROUP_SIZE == 0) {
                    sbuilder.append("\n\n");
                }
            }
        }

        sbuilder.append(CLOSE_SPOILER);
    }

    private void generateCommandsList(StringBuilder sbuilder, List<VillageAssignment> assignmentsList, int speed) {
        String departureTime, possibleNewLine, assignment, domain;
        int counter = 0, departureId, destinationId;

        sbuilder.append(OPEN_SPOILER);

        for (VillageAssignment a : assignmentsList) {
            departureTime = getDepartureTime(a.getSquaredDistance(), speed);

            possibleNewLine = nextDay ? "\n\n\n" : "";
            assignment = a.toString();
            domain = world.getDomain();
            departureId = a.getDeparture().getId();
            destinationId = a.getDestination().getId();

            sbuilder.append( String.format("%s%d. %s %s\n", possibleNewLine, ++counter, departureTime, assignment) );

            sbuilder.append( String.format("[url=%s/game.php?village=%d&screen=place&target=%d]%s[/url]\n", domain, departureId, destinationId, EXECUTION_TEXT) );
        }

        sbuilder.append(CLOSE_SPOILER);
    }

    //TODO implement custom dateOfArrival
    private String getDepartureTime(double squaredDistance, int standardTimePerUnit) {
        Calendar dateOfArrival = new GregorianCalendar(2018, Calendar.MAY, 21);
        dateOfArrival.add(Calendar.HOUR, 8);

        double distance = Math.sqrt(squaredDistance);
        double unitSpeed = standardTimePerUnit / world.getSpeed();
        int travelTime = (int) ((distance * unitSpeed) * SECONDS_PER_MIN );

        dateOfArrival.add(Calendar.SECOND, -travelTime);

        int dayOfMonth = dateOfArrival.get(Calendar.DAY_OF_MONTH);

        if (dayOfMonth != previousDay) {
            nextDay = true;
        } else {
            nextDay = false;
        }
        previousDay = dayOfMonth;

        return new SimpleDateFormat("dd.MM | HH:mm:ss").format(dateOfArrival.getTime());
    }

    private void sortAssignmentsList(List<VillageAssignment> list) {
        list.sort(Comparator.comparing(VillageAssignment::getSquaredDistance));
        Collections.reverse(list);
    }

    private HashMap<Integer, List<Village>> sortRequirements(HashMap<Village, Integer> map) {
        HashMap<Integer, List<Village>> sortedMap = new HashMap<>();

        int value;
        List<Village> vList;
        for (Village vil : map.keySet()) {
            value = map.get(vil);
            if (sortedMap.get(value) == null) {
                vList = new LinkedList<>();
                vList.add(vil);
                sortedMap.put(value, vList);
            } else {
                sortedMap.get(value).add(vil);
            }
        }

        return sortedMap;
    }

    private int getUnitSpeed(Unit type) {
        int speed;
        switch (type) {
            case RAM:
                speed = 30;
                break;
            case NOBLE:
                speed = 35;
                break;
            default:
                speed = 30;
                break;
        }
        return speed;
    }
}

File: src/main/kotlin/io/github/stiv3ns/twactionorganizer/core/utils/PMFormatter.kt
package io.github.stiv3ns.twactionorganizer.core.utils

import io.github.stiv3ns.twactionorganizer.core.Player
import io.github.stiv3ns.twactionorganizer.core.Assignment
import io.github.stiv3ns.twactionorganizer.core.World
import io.github.stiv3ns.twactionorganizer.core.villages.Village
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.*
import kotlin.math.sqrt

class PMFormatter(private val world: World, private val dateOfArrival: LocalDateTime) {

    private val bundle = ResourceBundle.getBundle("localization/pmformatter")

    private val OPEN_SPOILER        by lazy { bundle.getString("OPEN_SPOILER") }
    private val CLOSE_SPOILER       by lazy { bundle.getString("CLOSE_SPOILER") }
    private val REQUIREMENTS_HEADER by lazy { bundle.getString("REQUIREMENTS_HEADER") }
    private val NOBLE_HEADER        by lazy { bundle.getString("NOBLE_HEADER") }
    private val OFF_HEADER          by lazy { bundle.getString("OFF_HEADER") }
    private val FAKE_HEADER         by lazy { bundle.getString("FAKE_HEADER") }
    private val FAKENOBLE_HEADER    by lazy { bundle.getString("FAKENOBLE_HEADER") }
    private val EXECUTION_TEXT      by lazy { bundle.getString("EXECUTION_TEXT") }

    private val GROUP_SIZE = 5

    private enum class TroopType {
        RAM,
        NOBLE
    }

    private fun troopTravelTime(type: TroopType) = when (type) {
        TroopType.RAM -> 30
        TroopType.NOBLE -> 35
    } / world.speed

    fun getFormattedMsgFor(player: Player): String {
        val msg = StringBuilder()

        if (player.nobleAssignments.isNotEmpty() || player.fakeNobleAssignments.isNotEmpty()) {
            msg.append(REQUIREMENTS_HEADER)
            appendNobleRequirements(msg, player)
        }

        if (player.nobleAssignments.isNotEmpty()) {
            msg.append(NOBLE_HEADER)
            appendAssignments(msg, player.nobleAssignments, TroopType.NOBLE)
        }

        if (player.offAssignments.isNotEmpty()) {
            msg.append(OFF_HEADER)
            appendAssignments(msg, player.offAssignments, TroopType.RAM)
        }

        if (player.fakeNobleAssignments.isNotEmpty()) {
            msg.append(FAKENOBLE_HEADER)
            appendAssignments(msg, player.fakeNobleAssignments, TroopType.NOBLE)
        }

        if (player.fakeAssignments.isNotEmpty()) {
            msg.append(FAKE_HEADER)
            appendAssignments(msg, player.fakeAssignments, TroopType.RAM)
        }

        return msg.toString()
    }

    private fun appendNobleRequirements(msg: StringBuilder, player: Player) {
        val allNobleAssignments =
            player.nobleAssignments + player.fakeNobleAssignments

        val requirementsByVillage =
            allNobleAssignments.groupBy { it.departure }.mapValues { it.value.size }

        val sortedRequirements =
            requirementsByVillage.toList().sortedByDescending { (_, cnt) -> cnt }.toMap()

        msg.append(OPEN_SPOILER)

        var iteratorCounter = 0
        for ((village, numberOfRequiredNobles) in sortedRequirements) {
            msg.append("$numberOfRequiredNobles __ $village\n")

            if (++iteratorCounter % GROUP_SIZE == 0)
                msg.append("\n")
        }

        msg.append(CLOSE_SPOILER)
    }

    private fun appendAssignments(msg: StringBuilder, assignments: List<Assignment>, slowestTroop: TroopType) {
        msg.append(OPEN_SPOILER)

        val sortedAssignments = assignments.sortedByDescending { it.squaredDistance }
        var counter = 0
        var previousDayOfMonth = -1

        for (assignment in sortedAssignments) {
            val departureTime = designateDepartureTime(
                assignment.squaredDistance,
                slowestTroop,
                assignment.delayInMinutes
            )

            val verticalSpaceIfNextDay =
                if (previousDayOfMonth !in listOf(departureTime.dayOfMonth, -1))
                    "\n\n\n\n\n!"
                else ""

            val verticalSpaceIfNextGroup = if (counter % GROUP_SIZE == 0) "\n" else ""

            msg.append("$verticalSpaceIfNextDay ${++counter}. ${formatDepartureTime(departureTime)} $assignment \n")
                .append("[url=${world.domain}/game.php")
                .append("?village=${assignment.departure.id ?: 0}")
                .append("&screen=place&target=${assignment.destination.id ?: 0}]")
                .append("$EXECUTION_TEXT[/url]\n")
                .append(verticalSpaceIfNextGroup)

            previousDayOfMonth = departureTime.dayOfMonth
        }

        msg.append(CLOSE_SPOILER)
    }

    private fun designateDepartureTime(squaredDistance: Int, slowestTroop: TroopType, delayInMinutes: Long): LocalDateTime {
        val actualDistance = sqrt(squaredDistance.toDouble())
        val travelTime = actualDistance * troopTravelTime(slowestTroop)

        return dateOfArrival.minusMinutes(travelTime.toLong()).plusMinutes(delayInMinutes)
    }

    private fun formatDepartureTime(departureTime: LocalDateTime): String =
        departureTime.format(DateTimeFormatter.ofPattern("dd/MM/YY | HH:mm:ss"))
}